### ISSUE TITLE
Reduce contention from spawning task in waitEndCount

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#12247]
We convert coforalls like:

```chpl 
coforall loc in Locales do on loc { body(); }
```

roughly into something like:

```chpl
var endCount: atomic int = Locales.size;
for loc in Locales {
  executeOnNb(loc, bodyWrap());
}
while (endCount.read() != 0) {
  chpl_task_yield();
}

proc bodyWrap() {
  body();
  endCount.sub(1);
}
```

Where the original task will spawn all other tasks and then spinwait until the `endCount` is 0. For local coforalls we use processor atomics so there's probably not a ton of overhead to this waiting, but for remote-coforalls we use network atomics when they are available. This means we end do a local nic-get in a loop. 

Under ugni this results in pci traffic and some overhead on the NIC itself, and also thrashes our remote memory lookup cache for tasks running on the same thread.

Benchmarks like RA do a lot of single word updates, and we yield while waiting for those updates to complete, which means the spawning task runs often, which can cause interference.

Possible solutions: 
 - long term it would be nice to have the spawning task participate in the coforall. This would eliminate a task creation, and prevent the spinwaiting from happening until at least part of the kernel was done. There are issues here with task stack space though, which would be particularly tricky for nested task creation.
 - We could spin on a processor atomic and have the last task to finish do an AM to update that atomic (we could also wait on a sync var to put the task to sleep): https://github.com/ronawho/chapel-archive/compare/put-spawning-task-to-sleep
 - We might be able to use a non-atomic peek() for network atomics. This would allow us to just read from main memory instead of going through the NIC and at least under ugni (since we don't used CACHED atomics) we could rely on that value being propagated to memory once there are no incoming updates. We do have to worry about torn values still so we would have to do a full read() after out peek() tells us the value is 0: https://github.com/ronawho/chapel-archive/compare/use-peek-endcount-wait
